code/game/objects/items/mop.dm

```diff
@@ -62,13 +63,13 @@
 		return
 
 	if(T)
-		user.visible_message(span_notice("[user] begins to clean \the [T] with [src]."), span_notice("You begin to clean \the [T] with [src]..."))
-		var/clean_speedies = 1
-		if(user.mind)
-			clean_speedies = user.mind.get_skill_modifier(/datum/skill/cleaning, SKILL_SPEED_MODIFIER)
-		if(do_after(user, mopspeed*clean_speedies, target = T))
-			to_chat(user, span_notice("You finish mopping."))
-			clean(T, user)
+		var/datum/component/cleaner/cleaner = GetComponent(/datum/component/cleaner)
+		if(!reagents.has_chemical_flag(REAGENT_CLEANS, 1)) //won't clean the turf nor give you experience without cleaning reagents
```

Comment:
Can't we just early return if there's no chemicals, rather than GetComponent and set to not give XP?

Reply:
Can't early return because you can still mop with chemicals that don't actually clean. So if you dip your mop in welding fuel for example, you still get the do_after, skill speed modifier, the cleaning message and it applies the welding fuel to the floor. Setting the washing strength and XP modifier was an easy way to cheat and use the same code without actually washing the tile. I can't really think of a way to avoid using GetComponent without having to duplicate some code that the cleaning component is already supposed to take care of.

Reply:
You can make it an arg for COMSIG_START_CLEANING, maybe. It would be preferred over GetComponent and setting exp gain to 0, cleaning, then setting it back to 1.